core/common/src/main/java/alluxio/util/JvmPauseMonitor.java

```diff
@@ -0,0 +1,213 @@
+import com.google.common.collect.Lists;
+import com.google.common.base.Stopwatch;
+
+import com.google.common.collect.Sets;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.lang.management.GarbageCollectorMXBean;
+import java.lang.management.ManagementFactory;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class to monitor JVM with a daemon thread, the thread sleep period of time
+ * and get the true time the sleep takes. If it is longer than it should be, the
+ * JVM has paused processing. Then log it into different level.
+ */
+@NotThreadSafe
+public final class JvmPauseMonitor {
+  private static final Log LOG = LogFactory.getLog(
+      JvmPauseMonitor.class);
+
+  /** The time to sleep. */
+  private final long mGcSleepIntervalMs;
+
+  /** Extra sleep time longer than this threshold, log WARN. */
+  private final long mWarnThresholdMs;
+
+  /** Extra sleep time longer than this threshold, log INFO. */
+  private final long mInfoThresholdMs;
+
+  /** Times extra sleep time exceed WARN. */
+  private long mWarnTimeExceededMS = 0;
+  /** Times extra sleep time exceed INFO. */
+  private long mInfoTimeExceededMS = 0;
+  /** Total extra sleep time. */
+  private long mTotalExtraTimeMs = 0;
+
+  private Thread mJvmMonitorThread;
+
+  /**
+   * Constructs JvmPauseMonitor.
+   */
+  public JvmPauseMonitor() {
+    mGcSleepIntervalMs = Configuration.getLong(PropertyKey.JVM_MONITOR_SLEEP_INTERVAL_MS);
+    mWarnThresholdMs = Configuration.getLong(PropertyKey.JVM_MONITOR_WARN_THRESHOLD_MS);
+    mInfoThresholdMs = Configuration.getLong(PropertyKey.JVM_MONITOR_INFO_THRESHOLD_MS);
+  }
+
+  /**
+   * Starts jvm monitor thread.
+   */
+  public void start() {
+    Preconditions.checkState(mJvmMonitorThread == null,
+        "JVM monitor thread already started");
+    mJvmMonitorThread = new Daemon(new Monitor());
+    mJvmMonitorThread.start();
+  }
+
+  /**
+   * Stops jvm monitor.
+   */
+  public void stop() {
+    Preconditions.checkState(mJvmMonitorThread != null,
+        "JVM monitor thread does not start");
+    mJvmMonitorThread.interrupt();
+    try {
+      mJvmMonitorThread.join();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+    reset();
+  }
+
+  /**
+   * Resets value of mJvmMonitorThread.
+   */
+  public void reset() {
+    mJvmMonitorThread = null;
+  }
+
+  /**
+   * @return true if thread started,false otherwise
+   */
+  public boolean isStarted() {
+    return mJvmMonitorThread != null;
+  }
+
+  /**
+   * @return time exceeds WARN threshold in milliseconds
+   */
+  public long getWarnTimeExceeded() {
+    return mWarnTimeExceededMS;
+  }
+
+  /**
+   * @return time exceeds INFO threshold in milliseconds
+   */
+  public long getInfoTimeExceeded() {
+    return mInfoTimeExceededMS;
+  }
```

Comment:
Instead of exposing these methods for testing, you can try validating the logger wrote the expected output.